File: cli/src/subnet.ts
import { spawnSync } from 'child_process';
import { isValidCidr } from '@/validators.js';

export interface CidrRange {
  start: number;
  end: number;
  mask: number;
}

export function parseCidr(cidr: string): CidrRange | null {
  if (!isValidCidr(cidr)) return null;
  const [addr, maskStr] = cidr.split('/');
  const mask = Number(maskStr);
  const parts = addr.split('.').map(Number);
  const ip = (parts[0] << 24) | (parts[1] << 16) | (parts[2] << 8) | parts[3];
  const maskBits = mask === 0 ? 0 : (~0 << (32 - mask)) >>> 0;
  const start = (ip & maskBits) >>> 0;
  const end = (start | (~maskBits >>> 0)) >>> 0;
  return { start, end, mask };
}

function toIp(n: number): string {
  return [(n >>> 24) & 0xff, (n >>> 16) & 0xff, (n >>> 8) & 0xff, n & 0xff].join('.');
}

function toCidr(start: number, mask: number): string {
  return `${toIp(start)}/${mask}`;
}

export function rangesOverlap(a: CidrRange, b: CidrRange): boolean {
  return a.start <= b.end && b.start <= a.end;
}

export function listUsedSubnets(): CidrRange[] {
  const ids = spawnSync('docker', ['network', 'ls', '--quiet'], { encoding: 'utf8' });
  if (ids.status !== 0) return [];
  const idList = String(ids.stdout ?? '')
    .split('\n')
    .map((s) => s.trim())
    .filter(Boolean);
  if (idList.length === 0) return [];
  const r = spawnSync(
    'docker',
    [
      'network',
      'inspect',
      '--format',
      '{{range .IPAM.Config}}{{.Subnet}}\n{{end}}',
      ...idList,
    ],
    { encoding: 'utf8' },
  );
  if (r.status !== 0) return [];
  const out: CidrRange[] = [];
  for (const line of String(r.stdout ?? '').split('\n')) {
    const s = line.trim();
    if (!s) continue;
    const c = parseCidr(s);
    if (c) out.push(c);
  }
  return out;
}

function* candidates(): IterableIterator<string> {
  for (let c = 0; c <= 255; c++) {
    for (let d = 0; d <= 240; d += 16) yield `172.25.${c}.${d}/28`;
  }
  for (let b = 26; b <= 254; b++) yield `172.${b}.0.0/28`;
  for (let b = 0; b <= 254; b++) yield `10.${b}.0.0/28`;
}

export function findFreeSubnet(preferred: string, used?: CidrRange[]): string {
  const usedRanges = used ?? listUsedSubnets();
  const pref = parseCidr(preferred);
  if (pref && !usedRanges.some((u) => rangesOverlap(pref, u))) return preferred;
  for (const cand of candidates()) {
    const c = parseCidr(cand);
    if (!c) continue;
    if (cand === preferred) continue;
    if (!usedRanges.some((u) => rangesOverlap(c, u))) return cand;
  }
  return preferred;
}

export function subnetConflict(cidr: string, used?: CidrRange[]): CidrRange | null {
  const c = parseCidr(cidr);
  if (!c) return null;
  const list = used ?? listUsedSubnets();
  for (const u of list) {
    if (rangesOverlap(c, u)) return u;
  }
  return null;
}

export function formatCidr(r: CidrRange): string {
  return toCidr(r.start, r.mask);
}

export function nthHost(cidr: string, n: number): string | null {
  const r = parseCidr(cidr);
  if (!r) return null;
  const ip = (r.start + n) >>> 0;
  if (ip <= r.start || ip >= r.end) return null;
  return toIp(ip);
}

export function lastHost(cidr: string): string | null {
  const r = parseCidr(cidr);
  if (!r) return null;
  const ip = (r.end - 1) >>> 0;
  if (ip <= r.start || ip >= r.end) return null;
  return toIp(ip);
}

File: cli-go/internal/domain/subnet.go
package domain

import (
	"fmt"
	"strings"
)

type CidrRange struct {
	Start uint32
	End   uint32
	Mask  uint32
}

func ParseCidr(cidr string) (*CidrRange, bool) {
	if !IsValidCidr(cidr) {
		return nil, false
	}
	parts := strings.SplitN(cidr, "/", 2)
	var mask uint32
	fmt.Sscanf(parts[1], "%d", &mask)
	octets := strings.Split(parts[0], ".")
	var ip uint32
	for _, o := range octets {
		var n uint32
		fmt.Sscanf(o, "%d", &n)
		ip = (ip << 8) | n
	}
	var maskBits uint32
	if mask == 0 {
		maskBits = 0
	} else {
		maskBits = ^uint32(0) << (32 - mask)
	}
	start := ip & maskBits
	end := start | (^maskBits)
	return &CidrRange{Start: start, End: end, Mask: mask}, true
}

func toIP(n uint32) string {
	return fmt.Sprintf("%d.%d.%d.%d",
		(n>>24)&0xff,
		(n>>16)&0xff,
		(n>>8)&0xff,
		n&0xff,
	)
}

func RangesOverlap(a, b CidrRange) bool {
	return a.Start <= b.End && b.Start <= a.End
}

func FindFreeSubnet(preferred string, usedSubnets []CidrRange) string {
	pref, ok := ParseCidr(preferred)
	if ok {
		overlaps := false
		for _, u := range usedSubnets {
			if RangesOverlap(*pref, u) {
				overlaps = true
				break
			}
		}
		if !overlaps {
			return preferred
		}
	}
	for _, cand := range subnetCandidates() {
		if cand == preferred {
			continue
		}
		c, ok := ParseCidr(cand)
		if !ok {
			continue
		}
		overlaps := false
		for _, u := range usedSubnets {
			if RangesOverlap(*c, u) {
				overlaps = true
				break
			}
		}
		if !overlaps {
			return cand
		}
	}
	return preferred
}

func SubnetConflict(cidr string, usedSubnets []CidrRange) *CidrRange {
	c, ok := ParseCidr(cidr)
	if !ok {
		return nil
	}
	for _, u := range usedSubnets {
		if RangesOverlap(*c, u) {
			return &u
		}
	}
	return nil
}

func FormatCidr(r CidrRange) string {
	return fmt.Sprintf("%s/%d", toIP(r.Start), r.Mask)
}

func NthHost(cidr string, n int) (string, bool) {
	r, ok := ParseCidr(cidr)
	if !ok {
		return "", false
	}
	ip := r.Start + uint32(n)
	if ip <= r.Start || ip >= r.End {
		return "", false
	}
	return toIP(ip), true
}

func LastHost(cidr string) (string, bool) {
	r, ok := ParseCidr(cidr)
	if !ok {
		return "", false
	}
	ip := r.End - 1
	if ip <= r.Start || ip >= r.End {
		return "", false
	}
	return toIP(ip), true
}

func subnetCandidates() []string {
	var candidates []string
	for c := 0; c <= 255; c++ {
		for d := 0; d <= 240; d += 16 {
			candidates = append(candidates, fmt.Sprintf("172.25.%d.%d/28", c, d))
		}
	}
	for b := 26; b <= 254; b++ {
		candidates = append(candidates, fmt.Sprintf("172.%d.0.0/28", b))
	}
	for b := 0; b <= 254; b++ {
		candidates = append(candidates, fmt.Sprintf("10.%d.0.0/28", b))
	}
	return candidates
}
